presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java

```diff
@@ -247,98 +248,54 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
     {
         InformationSchemaTableHandle table = (InformationSchemaTableHandle) handle;
 
-        if (!table.getPrefixes().equals(defaultPrefixes())) {
+        // The table is catalog-wise and doesn't have a schema/table name filter.
+        if (!table.getSchemaPrefixes().isPresent()) {
             return Optional.empty();
         }
 
-        Set<QualifiedTablePrefix> prefixes = getPrefixes(session, table, constraint);
-
-        table = new InformationSchemaTableHandle(table.getCatalogName(), table.getSchemaName(), table.getTableName(), prefixes);
-
-        return Optional.of(new ConstraintApplicationResult<>(table, constraint.getSummary()));
-    }
-
-    private Set<QualifiedTablePrefix> defaultPrefixes()
-    {
-        return ImmutableSet.of(new QualifiedTablePrefix(catalogName));
-    }
-
-    private Set<QualifiedTablePrefix> getPrefixes(ConnectorSession session, InformationSchemaTableHandle table, Constraint constraint)
-    {
-        InformationSchemaTableHandle handle = checkTableHandle(table);
+        // TODO: constraint may have two levels of filters: schema-level and table-level. It also has two types of filters: discrete (=) and consecutive (>, <, LIKE, etc.).
+        //  Discrete and consecutive schema-level filters are supported;
+        //  discrete tabel-level filters are supported;
+        //  consecutive table-level filters are not yet supported.
+        Set<QualifiedTablePrefix> newSchemaPrefixes = applyFilterToSchemaPrefixes(table.getSchemaPrefixes().get(), constraint);
+        Optional<Set<String>> newTableNameDomain = filterString(constraint.getSummary(), TABLE_NAME_COLUMN_HANDLE);
 
-        if (constraint.getSummary().isNone()) {
-            return ImmutableSet.of();
+        // Neither the schema prefixes nor the table name domain changes.
+        if (newSchemaPrefixes.size() == table.getSchemaPrefixes().get().size()
+                && (!newTableNameDomain.isPresent() || (table.getTableNameDomain().isPresent() && table.getTableNameDomain().get().size() == newTableNameDomain.get().size()))) {
+            return Optional.empty();
         }
 
-        Set<QualifiedTablePrefix> prefixes = calculatePrefixesWithSchemaName(session, constraint.getSummary(), constraint.predicate());
-        if (isTablesEnumeratingTable(handle.getSchemaTableName())) {
-            Set<QualifiedTablePrefix> tablePrefixes = calculatePrefixesWithTableName(session, prefixes, constraint.getSummary(), constraint.predicate());
-            // in case of high number of prefixes it is better to populate all data and then filter
-            if (tablePrefixes.size() <= MAX_PREFIXES_COUNT) {
```

Comment:
@kokosing, without `MAX_PREFIXES_COUNT`, will certain information_schema queries be (significantly) slower?

Reply:
I think we need to understand what exact calls we are doing to metastore in both cases to see what it actually changes.

It could be verified using JMX connector.